src/uct/tcp/tcp_ep.c

```diff
@@ -1038,6 +1264,65 @@ ucs_status_t uct_tcp_ep_am_zcopy(uct_ep_h uct_ep, uint8_t am_id, const void *hea
     return status;
 }
 
+ucs_status_t uct_tcp_ep_put_zcopy(uct_ep_h uct_ep, const uct_iov_t *iov,
+                                  size_t iovcnt, uint64_t remote_addr,
+                                  uct_rkey_t rkey, uct_completion_t *comp)
+{
+    uct_tcp_ep_t *ep             = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface       = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr        = NULL;
+    uct_tcp_ep_put_req_t put_req = { 0 };
+    uct_tcp_ep_zcopy_tx_t *ctx;
+    ucs_status_t status;
+
+    status = uct_tcp_ep_prepare_zcopy(iface, ep, 0, &put_req, sizeof(put_req),
+                                      iov, iovcnt, "put_zcopy",
+                                      UCT_TCP_EP_PUT_ZCOPY_MAX -
+                                      sizeof(uct_tcp_am_hdr_t),
+                                      /* Set a payload length directly to the
+                                       * TX length, since PUT Zcopy doesn't
+                                       * set the payload length to TCP AM hdr */
+                                      &ep->tx.length, &hdr);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ctx            = ucs_derived_of(hdr, uct_tcp_ep_zcopy_tx_t);
+    hdr->am_id     = UCT_TCP_EP_PUT_REQ_AM_ID;
+    hdr->length    = sizeof(put_req);
+    put_req.addr   = remote_addr;
+    put_req.length = ep->tx.length;
+
+    status = uct_tcp_ep_am_sendv(iface, ep, 0, hdr, UCT_TCP_EP_PUT_ZCOPY_MAX,
+                                 &put_req, ctx->iov, ctx->iov_cnt);
+    if (ucs_unlikely((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS))) {
+        goto out;
+    }
+
+    UCT_TL_EP_STAT_OP(&ep->super, PUT, ZCOPY, put_req.length);
+
+    /* Increment ep::outstanding in order to ensure returning
+     * UCS_ERR_NO_RESOURCE from flush functions and do progressing.
+     * It has to be decremented upon PUT ACK message receiving */
+    uct_tcp_ep_outstanding_inc(ep);
+
+    if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
+        /* Since EP has to wait for an acknowledgment for this PUT
+         * operation, pass NULL instead of the user's completion
+         * object and always return UCS_INPROGRESS. The completion
+         * will be updated upon receiving the PUT acknowledgment */
+        uct_tcp_ep_set_outstanding_zcopy(iface, ep, ctx, &put_req,
+                                         sizeof(put_req), comp);
+        return UCS_INPROGRESS;
+    }
+
+    status = UCS_OK;
```

Comment:
will we return OK in case of `uct_tcp_ep_am_sendv` returns `NO_PROGESS`?

Reply:
We return `UCS_INPROGRESS`, when `UCS_ERR_NO_PROGRESS` is returned.
Added `assert()`